File: jwrap2/jw/src/main/java/global/CborUtil.java
package global;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.RandomAccessFile;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import com.upokecenter.cbor.CBORObject;
import com.upokecenter.cbor.CBORType;

public class CborUtil {


	public static void PutArrayToFileEnd(String filePath, CBORObject list) throws Exception {
		try (RandomAccessFile file = new RandomAccessFile(filePath, "rw")) {
			long offset = file.length();
			file.seek(offset);
			byte[] cborBytes;
			try (ByteArrayOutputStream stream = new ByteArrayOutputStream()) {
				CBORObject.Write(list.size(), stream);
				for (int i = 0; i < list.size(); i++) {
					CBORObject.Write(list.get(i), stream);
				}
				cborBytes = stream.toByteArray();
			}
			file.write(cborBytes);
			file.write(new byte[] { 0 });
			String offsetString = "" + offset;
			byte[] offsetBytes = offsetString.getBytes("UTF-8");
			file.write(offsetBytes);
			file.close();
		}
	}

	public static void PutOneToFileEnd(String filePath, CBORObject one) throws Exception {
		try (RandomAccessFile file = new RandomAccessFile(filePath, "rw")) {
			long offset = file.length();
			file.seek(offset);
			byte[] cborBytes;
			try (ByteArrayOutputStream stream = new ByteArrayOutputStream()) {
				CBORObject.Write(1, stream);
				CBORObject.Write(one, stream);
				cborBytes = stream.toByteArray();
			}
			file.write(cborBytes);
			file.write(new byte[] { 0 });
			String offsetString = "" + offset;
			byte[] offsetBytes = offsetString.getBytes("UTF-8");
			file.write(offsetBytes);
			file.close();
		}
	}

	public static CBORObject GetArrayFromFileEnd(String filePath) throws Exception {
		try (RandomAccessFile file = new RandomAccessFile(filePath, "r")) {
			{
				long offset = file.length();
				while (offset > 0) {
					offset--;
					file.seek(offset);
					byte b = file.readByte();
					if (b == 0) {
						offset++;
						file.seek(offset);
						byte[] offsetBytes = new byte[(int) (file.length() - offset)];
						file.readFully(offsetBytes);
						String offsetString = new String(offsetBytes, "UTF-8");
						long start = Long.parseLong(offsetString);
						file.seek(start);
						byte[] bytes = new byte[(int) (file.length() - start)];
						file.readFully(bytes);
						CBORObject list = CBORObject.NewArray();
						try (ByteArrayInputStream stream = new ByteArrayInputStream(bytes)) {
							var count = CBORObject.Read(stream).AsInt32();
							if (count <= 0)
								return CBORObject.Undefined;
							for (int i = 0; i < count; i++) {
								var cbor = CBORObject.Read(stream);
								list.Add(cbor);
							}
						}
						return list;
					}
				}
				return CBORObject.Undefined;
			}
		}
	}

	public static CBORObject GetOneFromFileEnd(String filePath) throws Exception {
		try (RandomAccessFile file = new RandomAccessFile(filePath, "r")) {
			{
				long offset = file.length();
				while (offset > 0) {
					offset--;
					file.seek(offset);
					byte b = file.readByte();
					if (b == 0) {
						offset++;
						file.seek(offset);
						byte[] offsetBytes = new byte[(int) (file.length() - offset)];
						file.readFully(offsetBytes);
						String offsetString = new String(offsetBytes, "UTF-8");
						long start = Long.parseLong(offsetString);
						file.seek(start);
						byte[] bytes = new byte[(int) (file.length() - start)];
						file.readFully(bytes);
						CBORObject list = CBORObject.NewArray();
						try (ByteArrayInputStream stream = new ByteArrayInputStream(bytes)) {
							var count = CBORObject.Read(stream).AsInt32();
							if (count <= 0)
								return CBORObject.Undefined;
							var cbor = CBORObject.Read(stream);
							return cbor;
						}
					}
				}
				return CBORObject.Undefined;
			}
		}
	}

}

File: jwrap2/CborUtil.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;
using PeterO.Cbor;
using System;
using System.IO;
using System.IO.Compression;

public class CborUtil
{
    public static void PutArrayToFileEnd(string filePath, CBORObject list)
    {
        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Write))
        {
            long offset = stream.Length;
            stream.Seek(offset, SeekOrigin.Begin);
            CBORObject.Write(list.Count, stream);
            for (int i = 0; i < list.Count; i++)
            {
                CBORObject.Write(list[i], stream);
            }

            stream.WriteByte(0);
            string offsetString = $"{offset}";
            byte[] offsetBytes = Encoding.UTF8.GetBytes(offsetString);
            stream.Write(offsetBytes, 0, offsetBytes.Length);
        }
    }

    public static void PutOneToFileEnd(string filePath, CBORObject one)
    {
        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Write))
        {
            long offset = stream.Length;
            stream.Seek(offset, SeekOrigin.Begin);
            CBORObject.Write(1, stream);
            CBORObject.Write(one, stream);
            stream.WriteByte(0);
            string offsetString = $"{offset}";
            byte[] offsetBytes = Encoding.UTF8.GetBytes(offsetString);
            stream.Write(offsetBytes, 0, offsetBytes.Length);
        }
    }

    public static CBORObject GetArrayFromFileEnd(string filePath)
    {
        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
        {
            long offset = stream.Length;
            while (offset > 0)
            {
                offset--;
                stream.Seek(offset, SeekOrigin.Begin);
                int b = stream.ReadByte();
                if (b == 0)
                {
                    offset++;
                    stream.Seek(offset, SeekOrigin.Begin);
                    byte[] offsetBytes = new byte[stream.Length - offset];
                    stream.Read(offsetBytes, 0, offsetBytes.Length);
                    string offsetString = Encoding.UTF8.GetString(offsetBytes);
                    long start = long.Parse(offsetString);
                    stream.Seek(start, SeekOrigin.Begin);
                    var count = CBORObject.Read(stream).AsInt32();
                    CBORObject list = CBORObject.NewArray();
                    for (int i = 0; i < count; i++)
                    {
                        list.Add(CBORObject.Read(stream));
                    }

                    return list;
                }
            }

            return CBORObject.Undefined;
        }
    }

    public static CBORObject GetOneFromFileEnd(string filePath)
    {
        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
        {
            long offset = stream.Length;
            while (offset > 0)
            {
                offset--;
                stream.Seek(offset, SeekOrigin.Begin);
                int b = stream.ReadByte();
                if (b == 0)
                {
                    offset++;
                    stream.Seek(offset, SeekOrigin.Begin);
                    byte[] offsetBytes = new byte[stream.Length - offset];
                    stream.Read(offsetBytes, 0, offsetBytes.Length);
                    string offsetString = Encoding.UTF8.GetString(offsetBytes);
                    long start = long.Parse(offsetString);
                    stream.Seek(start, SeekOrigin.Begin);
                    var count = CBORObject.Read(stream).AsInt32();
                    if (count <= 0) return CBORObject.Undefined;
                    CBORObject list = CBORObject.NewArray();
                    return CBORObject.Read(stream);
                }
            }

            return CBORObject.Undefined;
        }
    }

}